Luca/fix raw predict issue 635

lucagiovagnoli opened the pull request:
**Code change**
This is a one line change that fixes issue #635 where rawPredict should be `(-m, m)`.
(also found another small bug where tree_limit wasn't being serialized in the `store`)

**Added tests**
I noticed there were no tests for mleap-xgboost-runtime so I added a bunch of them using the same "agaricus" dataset from `dmlc-xgboost`:

1. Booster and mleap XGBoostClassifier results match (tests for `(-m, m)`)
2. XGBoostClassification works before and after serialization.
3. XGBoostRuntime schema assertions based on the NodeShape


@talalryz @voganrc @ancasarb

ancasarb commented on mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/testing/BundleSerializationUtils.scala:
```diff
@@ -0,0 +1,35 @@
+  def serializeModelToMleapBundle(transformer: Transformer): File = {
+    import ml.combust.mleap.runtime.MleapSupport._
+
+    new File("/tmp/mleap/spark-parity").mkdirs()
```
perhaps "/tmp/mleap/xgboost-runtime-parity" as this doesn't test anything spark related?

lucagiovagnoli replied:
Oh yes, leftover from copy-pasting from the spark tests. Let me update it

ancasarb commented on project/Dependencies.scala:
```diff
@@ -114,7 +115,7 @@ object Dependencies {
   val tensorflow = l ++= tensorflowDeps ++ Seq(Test.scalaTest)
 
-  val xgboostRuntime = l ++= Seq(xgboostDep) ++ Seq(Test.scalaTest)
+  val xgboostRuntime = l ++= Seq(xgboostDep) ++ Test.spark ++ Seq(Test.scalaTest)
```
are you sure Provided.spark isn't enough? it seems like that for the other parity tests?

lucagiovagnoli replied:
So I actually meant to get your input on these dependency changes. 
Both Provided.spark or Test.spark achieve the same thing here because I am using spark things only in tests.

Now, I think that putting those dependencies with a `test` <scope> makes the build faster, since the dependencies are only brought in when running tests, not when building. I read online:

- test: 
> Dependencies with maven dependency scope test are not needed to build and run the project. They are needed to compile and run the unit tests.

- provided:
> Maven dependency scope provided is used during build and test the project. 

so I assume a test scope is slightly lighter than provided? Again, the xgboost-runtime code has no dependencies on spark stuff, but its tests do

ancasarb replied:
That makes sense, thank you!

ancasarb commented on project/MleapProject.scala:
```diff
@@ -110,7 +110,9 @@ object MleapProject {
-    dependencies = Seq(runtime)
+    dependencies = Seq(
+      runtime,
+      sparkTestkit % "test")
```
where do we use sparkTestkit?

lucagiovagnoli replied:
If I don't bring that in, I cannot use:
`import org.apache.spark.sql.mleap.TypeConverters`

 Do you think it is ok to use that in tests? (%"test" should not be included in the build)

ancasarb replied:
I think that comes from mleap-spark-base, but spark-testkit should be ok too.

ancasarb commented on mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/testing/CachedDatasetUtils.scala:
```diff
@@ -0,0 +1,50 @@
+    // Use Spark utils to load libsvm from disk
+    val spark = SparkSession.builder()
+      .master("local[2]")
+      .appName("XGBoostRuntimeClassificationModelParitySpec")
```
is this used for both tests?

lucagiovagnoli replied:
Yes. I'll change it to this.getClass.getName

talalryz commented on mleap-xgboost-runtime/src/main/scala/ml/combust/mleap/xgboost/runtime/bundle/ops/XGBoostClassificationOp.scala:
```diff
@@ -23,8 +23,10 @@ class XGBoostClassificationOp extends MleapOp[XGBoostClassification, XGBoostClas
+      model
+        .withValue("num_features", Value.int(obj.numFeatures))
+        .withValue("num_classes", Value.int(obj.numClasses))
+        .withValue("tree_limit", Value.int(obj.treeLimit))
```
Was treeLimit not being saved earlier, and this was a bug. If so, I think we should report it as a bug, and mention that bug here. It will help folks facing this issue!

lucagiovagnoli replied:
Serializing a model using this Op makes it unusable because the `load()` expects a `tree_limit` and it blows up when it's not there. But I am not sure if anyone stores a Runtime model? 

In any case, I opened an issue here https://github.com/combust/mleap/issues/639

lucagiovagnoli commented on mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/testing/BoosterUtils.scala:
```diff
@@ -0,0 +1,18 @@
+  final val xgboostParams: Map[String, Any] = Map(
+    "eta" -> 0.3,
+    "max_depth" -> 2,
+    "objective" -> "binary:logistic",
```
Yes, I think this should be already tested in the XGBoost library, no need to cross the boundary.

lucagiovagnoli commented:
Thanks all for reviewing! I added a multinomial classifier test and fixed all copy-paste leftovers. Let me know what you think about using spark % test for these tests.

lucagiovagnoli commented:
it seems that my new tests passed successfully (https://travis-ci.org/combust/mleap/jobs/649255264#L9688) but the python ones failed. I'll try rerun these

lucagiovagnoli commented:
I think this broke after the recent update of virtualenv to 20.0.0
See https://github.com/pypa/virtualenv/issues/1551

voganrc commented on mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/XGBoostClassificationModelParitySpec.scala:
```diff
@@ -0,0 +1,128 @@
+        val mleapRawPredictionColIndex = mleapResult.schema.indexOf("raw_prediction").get
+        val mleapProbabilityColIndex = mleapResult.schema.indexOf("probability").get
+
+        val singleRowDMatrix = r(1).asInstanceOf[SparseTensor[Double]].asXGB
```
Is the `1` here the index of the assembled features?

lucagiovagnoli replied:
Yes, I'll use a constant name to be clearer
* row(0) is the label
* row(1) is "features"

voganrc commented on mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/XGBoostClassificationModelParitySpec.scala:
```diff
@@ -0,0 +1,128 @@
+    val xgboostTransformer = createBoosterClassifier(trainBooster(xgboostParams, denseDataset))
+
+    val mleapBundle = serializeModelToMleapBundle(xgboostTransformer)
+    val deserializedTransformer: Transformer = loadMleapTransformerFromBundle(mleapBundle)
```
Why do we bother serializing/deserializing here if the test below asserts pre/post serialization consistency?

voganrc replied:
If serialization breaks, then both of these tests fail, but I think it would be better for only one to fail

lucagiovagnoli replied:
True, I'll remove that from here

voganrc commented on mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/XGBoostClassificationModelParitySpec.scala:
```diff
@@ -0,0 +1,128 @@
+
+  it("Results between the XGBoost4j booster and the MLeap Transformer are the same") {
+    val booster = trainBooster(xgboostParams, denseDataset)
+    val xgboostTransformer = createBoosterClassifier(trainBooster(xgboostParams, denseDataset))
```
So the transformer is created via another `trainBooster` call? How do we guarantee that it ends up building the same tree? Are the xgboostParams seeded?

lucagiovagnoli replied:
Originally I was re-using the booster object but then I thought that it must probably create the same tree by construction. It seems that it does - otherwise this test would/will fail

voganrc commented on mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/XGBoostClassificationModelParitySpec.scala:
```diff
@@ -0,0 +1,128 @@
+
+  it("Test XGBoostMultinomialClassificationModel results are the same pre and post serialization") {
+    val booster = trainBooster(xgboostParams, denseDataset)
+    val xgboostTransformer = createMultinomialBoosterClassifier(booster)
```
Is there a test for `MultinomialBoosterClassifier` parity with XGBoost4J anywhere?

lucagiovagnoli replied:
I added one using a mutli-label Iris dataset

voganrc commented on mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/testing/FloatingPointApproximations.scala:
```diff
@@ -0,0 +1,24 @@
+    if ((x - y).abs < precision) true else false
+  }
+
+  def almostEqualSequences(
```
Is it really necessary to implement this ourselves? Is there no implementation in a library somewhere?

lucagiovagnoli replied:
We looked into it last quarter and haven't found anything. This is the same code we currently use internally to check float equality to the 1e-7

talalryz commented on mleap-xgboost-runtime/src/test/scala/ml/combust/mleap/xgboost/runtime/testing/CachedDatasetUtils.scala:
```diff
@@ -0,0 +1,54 @@
+      .appName(s"${this.getClass.getName}")
+      .getOrCreate()
+
+    // This is the dataset used by dmls-XGBoost https://github.com/dmlc/xgboost/blob/master/demo/data/agaricus.txt.train
```
typo, dmls --> dmlc
